### PR TITLE
Add test for graph state circuit 

### LINKED ISSUE #4349: circuit.library.graph_state needs testing
The object `GraphState` (in `qiskit/circuit/library/graph_state.py`) needs testing.  [No line is currently coveraged](https://coveralls.io/builds/30439931/source?filename=qiskit/circuit/library/graph_state.py).

### PR DESCRIPTION
**Pull request by mantcep**
### Summary
Fixes #4349 

### Details and comments
Added a test by computing the expected unitary from adjacency matrix and comparing that against the circuit produced by `GraphState`. The expected unitary is computed using Eq. (9) from [this paper](https://arxiv.org/pdf/quant-ph/0307130.pdf). The test is general and supports an adjacency matrix for any number of qubits. Due to that, the computation for CZ gates is slightly more involved, as it is generalized to any number qubits and control-target connections. That is done by checking whether the corresponding state has both the control and target in state |1> and assigning -1 to corresponding diagonal elements of an identity matrix.

### CODE REVIEW

**Comment by Cryoris** [comment]
Hi @mantcep,

looking into this with @ajavadia we came up with a nice test, that does not require computing the exponentially large unitary and checks the graph state from another point-of-view than the circuit gates. In Section II.B of the paper, the graph state is defined via it's stabilizers (Eq. (7)). (These stabilizers exist because the graph's we're looking at are created with Clifford circuits.) 

In Eq. (6) it is defined how to compute these stabilizers. We can also get them from the circuit using 
```
from qiskit.quantum_info import Clifford
Clifford(circuit).stabilizers.pauli.to_label()
```
If that output matches the stabilizers from Eq. (6) the graph state is correct.

Here's a little snippet on how you could compute the stabilizers via Eq. (6):
```python
from typing import List

def get_stabilizers(A: List[List[int]]) -> List[str]:
    """Returns the stabilizers of the graph state constructed by A.
    
    Args:
        A: The adjacency matrix of the graph.
        
    Returns:
        The stabilizers.
    """
    stabilizers = []  # keep track of all stabilizers
    num_vertices = len(A)
    for vertex_a in range(num_vertices):  
        stabilizer = [None] * num_vertices  # Paulis must be put into right place 
        for vertex_b in range(num_vertices):
            if vertex_a == vertex_b:   # self-connection --> 'X'
                stabilizer[vertex_a] = 'X' 
            elif A[vertex_a][vertex_b] != 0:  # vertices connected --> 'Z'
                stabilizer[vertex_b] = 'Z'
            else:  # else --> 'I'
                stabilizer[vertex_b] = 'I'
                
        # need to reverse for Qiskit's tensoring order
        stabilizers.append(''.join(stabilizer)[::-1])
        
    return stabilizers
```
Then the stabilizers should be the same, e.g.
```python
# A = [[1, 1], [1, 1]]
# A = [[1, 1, 1], [1, 1, 0], [1, 0, 1]]
A = [[0, 1, 0, 0, 1], [1, 0, 1, 0, 0], [0, 1, 0, 1, 0], [0, 0, 1, 0, 1], [1, 0, 0, 1, 0]]
graph_state = GraphState(A)
stabilizers = Clifford(graph_state).stabilizer.pauli.to_labels()
ref_stabilizers = get_stabilizers(A)
print(stabilizers)
print(ref_stabilizers)
```
yields
```
['ZIIZX', 'IIZXZ', 'IZXZI', 'ZXZII', 'XZIIZ']
['ZIIZX', 'IIZXZ', 'IZXZI', 'ZXZII', 'XZIIZ']
```

I think it would be good if we could test the graph states with this, what do you think? If you have questions don't hesitate to ask!

**Comment by mantcep** [comment]
Hi @Cryoris,

I must say, this is an elegant solution to say the least! I am sorry that I have not figured this out myself and you had to spend time looking into this. Thank you for the additional background on this topic, I will make sure to study at least the basics of quantum error correction, as I do not know much about stabilizers and Clifford groups.

As for implementing the test itself, would you like me to prepare the code for the test and submit a change to the pull request? If so, would it make sense to include the nice function that you have written into one of the classes?

Finally, I have one more question I am interested in and sorry if it sounds silly. Do you know what is the performance of computing the stabilizers from a circuit as compared to the exponentially large unitary based on the gates?

Thanks again for the guidance!

**Comment by Cryoris** [comment]
> I must say, this is an elegant solution to say the least! I am sorry that I have not figured this out myself and you had to spend time looking into this. Thank you for the additional background on this topic, I will make sure to study at least the basics of quantum error correction, as I do not know much about stabilizers and Clifford groups.

Don't worry, we're all learning something new every day! I also didn't notice this when I opened this issue. 🙂  

> As for implementing the test itself, would you like me to prepare the code for the test and submit a change to the pull request? If so, would it make sense to include the nice function that you have written into one of the classes?

Yes you can just update your branch and then the PR will also be updated. I think you could include the code into the class, either as separate function or inside the `assertGraphStateIsCorrect`. 

> Finally, I have one more question I am interested in and sorry if it sounds silly. Do you know what is the performance of computing the stabilizers from a circuit as compared to the exponentially large unitary based on the gates?

I'm not quite sure @chriseclectic is the expert on this, but I think we can iterate over all gates and update the stabilizer table so the complexity should be of the order of gates in the Clifford circuit. Here's some more context on Clifford circuits and how they can be simulated efficiently: https://arxiv.org/pdf/quant-ph/0406196.pdf. This is also what's done in [Qiskit](https://qiskit.org/documentation/stubs/qiskit.quantum_info.Clifford.html).

**Comment by mantcep** [comment]
Hi again!

I have updated the branch based on the code you have provided 🙂 Let me know if I should be changing something to make it nicer.

I was thinking that maybe it would make sense to include the function you have provided in the `GraphState` class to have a specific method for getting the stabilizers. In the end, however, I figured that is probably not necessary, since we can always get the stabilizers using the other method you have shown – `Clifford(circuit).stabilizers.pauli.to_label()`. So I included your code inside `assertGraphStateIsCorrect`, since that seemed to make most sense to me.

Thank you again for the valuable input!

**Comment by mantcep** [comment]
Just one more comment I would like to add. In `hidden_linear_function.py`, there is the following piece of code that checks whether the adjacency matrix is symmetric:

https://github.com/Qiskit/qiskit-terra/blob/968000fd14d43935538fdebb3ef10194940633e1/qiskit/circuit/library/hidden_linear_function.py#L87-L89

I think it would make sense to add that to `GraphState` as well. Then another test could be added that checks whether this raises an error for a non-symmetric matrix, similar to the following test:

https://github.com/Qiskit/qiskit-terra/blob/968000fd14d43935538fdebb3ef10194940633e1/test/python/circuit/test_library.py#L182-L185

Let me know what you think about this, I would be more than glad to update my branch with this change as well 🙂

**Comment by ajavadia** [comment]
> I was thinking that maybe it would make sense to include the function you have provided in the GraphState class to have a specific method for getting the stabilizers. In the end, however, I figured that is probably not necessary, since we can always get the stabilizers using the other method you have shown – Clifford(circuit).stabilizers.pauli.to_label()

Another way of getting this can be `clifford.to_dict()` which gives you a dictionary of stabalizers like:
```
{'stabilizer': ['+XX', '+ZZ'], 'destabilizer': ['+IZ', '+XI']}
```

**Comment by mantcep** [comment]
Hi @ajavadia,

Thank you for the input, I will keep this in mind! For the purpose of testing the `GraphState`, though, I would assume that using `Clifford(graph_state).stabilizer.pauli.to_labels()` would be fine, right? Otherwise, is there any other input for my pull request or could we proceed with merging that?

**Comment by Cryoris** [comment]
@mantcep If you could add the symmetry test that would be great. 

I think either way of getting the stabilizers is fine, though the `to_dict` method is a bit more elegant since it requires less code. I'd say it's up to you!

**Comment by mantcep** [comment]
Hi @Cryoris,

Sure, I will add the check for symmetry and see which method works best for me to get the stabilizers! I think I can get that done quite quickly and will update the branch once it is ready.

Thank you again!

**Comment by mantcep** [comment]
Hi again @Cryoris,

I have added the symmetry check and the corresponding test. Both of these are based on similar implementation in `HiddenLinearFunction`.
As for getting the stabilizers, I have not made any changes to that. I was not very sure about implementing the phase we get from using the `to_dict()` method, which, as I understand, corresponds to the phase _p_ in the following:

https://github.com/Qiskit/qiskit-terra/blob/1425ebddce035edac2fc2f96dbde5f89c4da8ae5/qiskit/quantum_info/operators/symplectic/stabilizer_table.py#L31-L34

I hope you do not mind me taking this shortcut here for the purpose of this test.

Let me know what you think about this final commit and whether there are any other changes I should implement before we proceed with merging this 🙂

I am very thankful for your support on my very first contribution! I understand that this is time consuming for you, so I will try to be a lot more independent in the next one!

File: test/python/circuit/test_library.py [hunk 1]
```diff
@@ -185,6 +185,49 @@ def test_non_symmetric_raises(self):
...
+            # need to reverse for Qiskit's tensoring order
+            expected_stabilizers.append(''.join(stabilizer)[::-1])
+
+        self.assertListEqual(expected_stabilizers, stabilizers)
+
+    @data(
+        [[0, 1, 0, 0, 1], [1, 0, 1, 0, 0], [0, 1, 0, 1, 0], [0, 0, 1, 0, 1], [1, 0, 0, 1, 0]]
+    )
+    def test_graph_state(self, adjacency_matrix):
+        """Test if Graph State circuit produces correct matrix."""
```
Comment (Cryoris) [inline test/python/circuit/test_library.py:218]:
```suggestion
        """Verify the GraphState by checking if the circuit has the expected stabilizers."""
```

File: test/python/circuit/test_library.py [hunk 2]
```diff
@@ -185,6 +185,49 @@ def test_non_symmetric_raises(self):
             HiddenLinearFunction([[1, 1, 0], [1, 0, 1], [1, 1, 1]])
 
 
+@ddt
+class TestGraphStateLibrary(QiskitTestCase):
+    """Test the graph state circuit."""
+
+    def assertGraphStateIsCorrect(self, adjacency_matrix, graph_state):
+        """Assert that the Graph State circuit produces the correct matrix."""
```
Comment (Cryoris) [inline test/python/circuit/test_library.py:193]:
```suggestion
        """Check the stabilizers of the graph state against the expected stabilizers.
        
        Based on https://arxiv.org/pdf/quant-ph/0307130.pdf, Eq. (6).
        """
```

File: test/python/circuit/test_library.py [hunk 3]
```diff
@@ -185,6 +185,52 @@ def test_non_symmetric_raises(self):
             HiddenLinearFunction([[1, 1, 0], [1, 0, 1], [1, 1, 1]])
 
 
+@ddt
+class TestGraphStateLibrary(QiskitTestCase):
+    """Test the graph state circuit."""
+
+    def assertGraphStateIsCorrect(self, adjacency_matrix, graph_state):
+        """Check the stabilizers of the graph state against the expected stabilizers.
+
```
Comment (Cryoris) [inline test/python/circuit/test_library.py:194]:
```suggestion

```

**Comment by mantcep** [comment]
Thank you, @Cryoris! 🙂

**Pull request merged**